File: apps/web/types/multiplayer.ts
export interface AsyncChallenge {
  id: string;
  challenger_name: string;
  opponent_name: string;
  game_name: string;
  difficulty: string;
  challenger_score: number | null;
  opponent_score: number | null;
  status: string;
  expires_at: string;
  created_at: string;
}

export interface AsyncChallengeDetail extends AsyncChallenge {
  questions?: Record<string, unknown>[];
  result?: ChallengeResult;
}

export interface ChallengeResult {
  winner_id: string;
  result: 'win' | 'lose' | 'draw' | 'pending';
  xp_earned: number;
}

export interface CreateChallengeRequest {
  opponent_username: string;
  game_slug: string;
  difficulty: string;
}

export interface SubmitChallengeRequest {
  answers: { question_id: string; answer: string; time_taken: number }[];
  score: number;
}

export interface WordChainGame {
  id: string;
  opponent_name: string;
  is_vs_bot: boolean;
  bot_difficulty: string;
  current_word: string;
  my_turn: boolean;
  player1_score: number;
  player2_score: number;
  status: string;
  turn_expires_at: string;
  created_at: string;
}

export interface WordChainDetail extends WordChainGame {
  words_used: string[];
}

export interface WordSubmitResult {
  valid: boolean;
  score_delta: number;
  next_letter: string;
  bot_response?: string;
  game_over: boolean;
  winner_id?: string;
  message?: string;
}

export interface CreateWordChainRequest {
  opponent_username?: string;
  vs_bot?: boolean;
  bot_difficulty?: string;
}

export interface SubmitWordRequest {
  word: string;
}

export interface QuickMatchResult {
  match_id: string;
  room_id: string;
  opponent_id?: string;
  opponent_name?: string;
  status?: string;
  bot_options?: string[];
}

export interface QuickMatchBotResult {
  match_id: string;
  room_id: string;
  recommended: 'easy' | 'medium' | 'hard';
  bot_difficulty: 'easy' | 'medium' | 'hard';
  bot: { name: string; difficulty: string };
}

export interface RealtimeQuestion {
  id: string;
  text: string;
  options: string[];
  question_number: number;
  total: number;
}

export interface AnswerResult {
  player_id: string;
  is_correct: boolean;
  score_delta: number;
  new_score: number;
}

export interface OpponentProgress {
  player_id: string;
  questions_answered: number;
  current_score: number;
}

export interface GameOverResult {
  results: {
    player_id: string;
    username: string;
    score: number;
    correct: number;
    wrong: number;
    is_winner: boolean;
  }[];
  winner_id: string;
  xp_earned: number;
}

export interface RoomSettings {
  questions: number;
  category: string;
  difficulty: 'easy' | 'medium' | 'hard';
  timer: number;
  max_players: number;
  allow_bots: boolean;
}

export interface RoomMember {
  id: string;
  username: string;
  level: number;
  is_host: boolean;
}

export interface RoomResponse {
  room_code: string;
  game_slug: string;
  host_id: string;
  members: RoomMember[];
  settings: RoomSettings;
  status: string;
  expires_at: string;
  game_name?: string;
}

export interface TournamentPlayer {
  id: string;
  user_id?: string;
  display_name: string;
  is_bot: boolean;
  seed: number;
  status: string;
  final_rank: number;
}

export interface TournamentMatch {
  id: string;
  round: number;
  position: number;
  player1: TournamentPlayer | null;
  player2: TournamentPlayer | null;
  winner_player_id?: string;
  player1_score: number;
  player2_score: number;
  status: string;
  room_id: string;
  ends_at?: string;
}

export interface Tournament {
  id: string;
  name: string;
  game_slug: string;
  difficulty: string;
  mode: 'quick' | 'open_daily' | 'open_weekly';
  invite_code: string;
  max_players: number;
  status: string;
  host_id: string;
  champion_id?: string;
  players: TournamentPlayer[];
  matches: TournamentMatch[];
  created_at: string;
  scheduled_start_at?: string;
  registration_ends_at?: string;
  current_round_ends_at?: string;
  started_at?: string;
  finished_at?: string;
}

export interface CreateTournamentRequest {
  name: string;
  difficulty: string;
  mode: 'quick' | 'open_daily' | 'open_weekly';
  max_players: number;
}

export interface ReportTournamentMatchRequest {
  winner_player_id: string;
  player1_score: number;
  player2_score: number;
}

export interface BattleshipCell {
  ship: boolean;
  hit: boolean;
  miss: boolean;
  revealed?: boolean;
}

export interface BattleshipQuestion {
  id: string;
  text: string;
  target: { row: number; col: number };
}

export interface BattleshipMatch {
  id: string;
  difficulty: 'easy' | 'medium' | 'hard';
  status: 'active' | 'finished';
  is_vs_bot: boolean;
  opponent_name: string;
  my_side: 'player1' | 'player2';
  current_turn: 'player1' | 'player2' | '';
  my_turn: boolean;
  my_score: number;
  opponent_score: number;
  winner_id?: string;
  winner_side?: 'player1' | 'player2';
  my_board: BattleshipCell[][];
  target_board: BattleshipCell[][];
  pending_question?: BattleshipQuestion;
  turn_expires_at?: string;
  pending_expires_at?: string;
  reveal_used: boolean;
  log: string[];
  created_at: string;
  updated_at: string;
  finished_at?: string;
}

export interface CreateBattleshipMatchRequest {
  opponent_username?: string;
  vs_bot: boolean;
  bot_difficulty?: 'easy' | 'medium' | 'hard';
  difficulty: 'easy' | 'medium' | 'hard';
}

File: services/api/internal/model/multiplayer.go
package model

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type MultiplayerMatch struct {
	ID          uuid.UUID  `gorm:"type:uuid;primaryKey" json:"id"`
	RoomCode    string     `gorm:"size:10" json:"room_code"`
	GameID      uuid.UUID  `gorm:"type:uuid;not null" json:"game_id"`
	MatchType   string     `gorm:"size:20;not null" json:"match_type"`
	Difficulty  string     `gorm:"size:10" json:"difficulty"`
	Status      string     `gorm:"size:15;default:'waiting'" json:"status"`
	WinnerID    *uuid.UUID `gorm:"type:uuid" json:"winner_id"`
	StartedAt   *time.Time `json:"started_at"`
	FinishedAt  *time.Time `json:"finished_at"`
	CreatedAt   time.Time  `json:"created_at"`
}

func (m *MultiplayerMatch) BeforeCreate(tx *gorm.DB) error {
	if m.ID == uuid.Nil {
		m.ID = uuid.New()
	}
	return nil
}

type MatchParticipant struct {
	ID              uuid.UUID  `gorm:"type:uuid;primaryKey" json:"id"`
	MatchID         uuid.UUID  `gorm:"type:uuid;not null" json:"match_id"`
	UserID          *uuid.UUID `gorm:"type:uuid" json:"user_id"`
	BotType         string     `gorm:"size:20" json:"bot_type"`
	BotDifficulty   string     `gorm:"size:10" json:"bot_difficulty"`
	BotName         string     `gorm:"size:50" json:"bot_name"`
	GhostID         *uuid.UUID `gorm:"type:uuid" json:"ghost_id"`
	Score           int        `gorm:"default:0" json:"score"`
	CorrectAnswers  int        `gorm:"default:0" json:"correct_answers"`
	WrongAnswers    int        `gorm:"default:0" json:"wrong_answers"`
	XPEarned        int        `gorm:"default:0" json:"xp_earned"`
	Rank            int        `json:"rank"`
	IsWinner        bool       `gorm:"default:false" json:"is_winner"`
	JoinedAt        time.Time  `json:"joined_at"`
	FinishedAt      *time.Time `json:"finished_at"`
}

func (mp *MatchParticipant) BeforeCreate(tx *gorm.DB) error {
	if mp.ID == uuid.Nil {
		mp.ID = uuid.New()
	}
	return nil
}

type GhostReplay struct {
	ID             uuid.UUID       `gorm:"type:uuid;primaryKey" json:"id"`
	UserID         uuid.UUID       `gorm:"type:uuid;not null" json:"user_id"`
	GameID         uuid.UUID       `gorm:"type:uuid;not null" json:"game_id"`
	Difficulty     string          `gorm:"size:10" json:"difficulty"`
	Score          int             `json:"score"`
	TotalQuestions int             `json:"total_questions"`
	CorrectAnswers int             `json:"correct_answers"`
	WrongAnswers   int             `json:"wrong_answers"`
	Duration       int             `json:"duration"`
	EventsJSON     json.RawMessage `gorm:"type:jsonb;not null" json:"events_json"`
	IsActive       bool            `gorm:"default:true" json:"is_active"`
	CreatedAt      time.Time       `json:"created_at"`
}

func (g *GhostReplay) BeforeCreate(tx *gorm.DB) error {
	if g.ID == uuid.Nil {
		g.ID = uuid.New()
	}
	return nil
}

type GhostEvent struct {
	Timestamp  time.Duration `json:"timestamp"`
	EventType  string        `json:"event_type"`
	QuestionID string        `json:"question_id"`
	IsCorrect  bool          `json:"is_correct"`
	TimeTaken  time.Duration `json:"time_taken"`
}

type AsyncChallenge struct {
	ID                uuid.UUID  `gorm:"type:uuid;primaryKey" json:"id"`
	ChallengerID      uuid.UUID  `gorm:"type:uuid;not null" json:"challenger_id"`
	OpponentID        uuid.UUID  `gorm:"type:uuid;not null" json:"opponent_id"`
	GameID            uuid.UUID  `gorm:"type:uuid;not null" json:"game_id"`
	Difficulty        string     `gorm:"size:10" json:"difficulty"`
	QuestionsJSON     json.RawMessage `gorm:"type:jsonb;not null" json:"questions_json"`
	AnswersJSON       json.RawMessage `gorm:"type:jsonb;not null" json:"answers_json"`
	ChallengerScore   *int       `json:"challenger_score"`
	OpponentScore     *int       `json:"opponent_score"`
	ChallengerDoneAt  *time.Time `json:"challenger_done_at"`
	OpponentDoneAt    *time.Time `json:"opponent_done_at"`
	WinnerID          *uuid.UUID `gorm:"type:uuid" json:"winner_id"`
	Status            string     `gorm:"size:15;default:'pending'" json:"status"`
	ExpiresAt         time.Time  `gorm:"not null" json:"expires_at"`
	CreatedAt         time.Time  `json:"created_at"`
	UpdatedAt         time.Time  `json:"updated_at"`
}

func (ac *AsyncChallenge) BeforeCreate(tx *gorm.DB) error {
	if ac.ID == uuid.Nil {
		ac.ID = uuid.New()
	}
	return nil
}

type WordChainGame struct {
	ID             uuid.UUID  `gorm:"type:uuid;primaryKey" json:"id"`
	Player1ID      uuid.UUID  `gorm:"type:uuid;not null" json:"player1_id"`
	Player2ID      *uuid.UUID `gorm:"type:uuid" json:"player2_id"`
	IsVsBot        bool       `gorm:"default:false" json:"is_vs_bot"`
	BotDifficulty  string     `gorm:"size:10" json:"bot_difficulty"`
	CurrentWord    string     `gorm:"size:50" json:"current_word"`
	CurrentTurn    *uuid.UUID `gorm:"type:uuid" json:"current_turn"`
	WordsUsed      []string   `gorm:"type:text[]" json:"words_used"`
	Player1Score   int        `gorm:"default:0" json:"player1_score"`
	Player2Score   int        `gorm:"default:0" json:"player2_score"`
	Status         string     `gorm:"size:15;default:'active'" json:"status"`
	TurnExpiresAt  *time.Time `json:"turn_expires_at"`
	CreatedAt      time.Time  `json:"created_at"`
	UpdatedAt      time.Time  `json:"updated_at"`
}

func (wc *WordChainGame) BeforeCreate(tx *gorm.DB) error {
	if wc.ID == uuid.Nil {
		wc.ID = uuid.New()
	}
	return nil
}
